### PR TITLE
exclude image-field from notebook controller reconciliation for all containers present in annotation

### PR DESCRIPTION
**Pull request by shalberd**
fixes #6829

[superseded by PR against v1.7-branch in PR-133](https://github.com/opendatahub-io/kubeflow/pull/133)
Looking for feedback before thinking of merging. AFIK, this should be pretty close to being mergeable, though.

seems to build alright locally, except unclear to me how to update the common/components, where reconcilehelper lies.
Got an error that is probably easy to resolve and dependent on the build chain:

```
#13 51.49 controllers/notebook_controller.go:213:58: too many arguments in call to "github.com/kubeflow/kubeflow/components/common/reconcilehelper".CopyStatefulSetFields                                                                                                     
#13 51.49 	have (*"k8s.io/api/apps/v1".StatefulSet, *"k8s.io/api/apps/v1".StatefulSet, bool, []string, logr.Logger)
#13 51.49 	want (*"k8s.io/api/apps/v1".StatefulSet, *"k8s.io/api/apps/v1".StatefulSet)
```

**that Problem was resolved by referencing our version of the reconcilehelper utilities-module under /components/common in notebook controlller go.mod:**

```
// use our version of kubeflow/components/common module for reconcilehelper utility differences related to image-field reconciliation and affected containers exclude
replace (
	github.com/kubeflow/kubeflow/components/common => ../common
)
```

Key idea:

Updating notebook controller logic and reconcilehelper util logic to take into account optional [Openshift image policy admission plug-in](https://docs.openshift.com/container-platform/4.10/openshift_images/triggering-updates-on-imagestream-changes.html#images-triggering-updates-imagestream-changes-kubernetes-about_triggering-updates-on-imagestream-changes) image-change trigger annotation with one or more containers and their related imagestreams (json array) that leads to container image-field update and to avoid overwriting the updated image-field value by the notebook controller.

[Discussed problem](https://github.com/openshift/openshift-apiserver/issues/339#issuecomment-1343192025) with @bparees , received input from @VaishnaviHire, member of [Open Data Hub (a RedHat project for Openshift) ](https://github.com/opendatahub-io) team. Also [discussed how to resolve the issue in Notebook reconciler with Ben](https://github.com/openshift/openshift-apiserver/issues/339#issuecomment-1357612359).

Context: Notebook yaml contains the (optional) metadata annotation

```
 annotations:
    image.openshift.io/triggers: '[{"from":{"kind":"ImageStreamTag","name":"s2i-generic-data-science-notebook:v0.0.5", "namespace":"odhsven"},"fieldPath":"spec.template.spec.containers[?(@.name==\"jupyter-nb-sven\")].image"}]'
```

referencing imagestream name and tag, imagestream namespace, and fieldPath reference to the image-value to be replaced by the image policy admission plug-in for a given container name. This annotation can reference one or more containers, in the form of a json array. This image-name resolve by the image policy admission plug-in happens in basic Kubernetes objects, not in the Notebook yaml itself, but in the StatefulSet derived from the Notebook yaml. The resolved image-name is set briefly after application of the StatefulSet to the server.

Or, referencing the openshift documentation:

_When one of the core Kubernetes resources contains both a pod template and this annotation, OpenShift Container Platform attempts to update the object by using the image currently associated with the image stream tag that is referenced by trigger. The update is performed against the fieldPath specified._

To ensure that image-fields referenced via this annotation in the StatefulSet are not overwritten by the kubeflow notebook controller reconcile process, we set image-fields-values before change and after change to be equal, thus leading to DeepEqual returning false even when the image-field value changes. 

Because the annotation is optional, checks are introduced in the code to ensure any custom logic is only done when the annotation is present.

Background: 

making image-field value independent of internal openshift registry and supporting ImageContentSourcePolicy for Openshift environments with a third-party Docker repo, like VMWare Harbor. This is archieved by using the imagestream abstraction. Referencing imagestream name and tag instead of image-digest-values directly. [Imagestreams are kind of an abstraction in openshift](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html/images/managing-image-streams).

See what happens to the image-field cause of the image policy admission plug-in doing its work:

- Without open shift registry:

`<image stream name>:<image stream tag>` references in image-field or annotation of a Kubernetes object (Deployment, StatefulSet, Pod etc.) always resolve to external image reference from imagestream spec.tags[tagname].from.name regardless of whether spec.tags[tagname].referencePolicy.type is set to Local or Source in image stream--> perfect for air gapped and ImageContentSourcePolicy use or if one just wants to pull directly from the external location when there is no internal openshift registry.
```
containers:
    - name: testissource
      image: >-
         quay.io/thoth-station/s2i-generic-data-science notebook@sha256:3f619c61501218f03d39d97541336dee024f446e64f3a47e2bc7e62cddeb2e58
```

- With openshift registry:

if  spec.tags[tagname].referencePolicy.type of the imagestream is set to Source then `<image stream name>:<image stream tag>` references in image-field or annotation of a Kubernetes object (Deployment, StatefulSet, Pod etc.) resolves to the remote repo location:

```
containers:
    - name: testissource
      image: >-
         quay.io/thoth-station/s2i-generic-data-science-notebook@sha256:3f619c61501218f03d39d97541336dee024f446e64f3a47e2bc7e62cddeb2e58
```

if  spec.tags[tagname].referencePolicy.type of the imagestream is set to Local then `<image stream name>:<image stream tag>` references in image-field or annotation of a Kubernetes object (Deployment, StatefulSet, Pod etc.) resolves to the namespace-specific location of the image in the open shift-internal registry:
```
containers
    - name: testisreflocal
      image: >-
        image-registry.openshift-image-registry.svc:5000/testis/jupyter-tensorflow-notebook@sha256:fc52e4fbc8c1c70dfa22dbfe6b0353f5165c507c125df4438fca6a3f31fe976e
```

### CODE REVIEW

**Comment by shalberd** [comment]
For testing, I will manually provide a Notebook CR as proposed in odh-dashboard PR https://github.com/opendatahub-io/odh-dashboard/pull/800 and see if the image-field gets updated accordingly.

File: components/notebook-controller/controllers/notebook_controller.go [hunk 1]
```diff
@@ -159,8 +204,12 @@ func (r *NotebookReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		log.Error(err, "error getting Statefulset")
 		return ctrl.Result{}, err
 	}
+
+        isImageChangeTriggerSet := metav1.HasAnnotation(instance.ObjectMeta, AnnotationImageChangeTrigger)
+        imageChangeTriggerReferencedContainerNames := getImageChangeTriggerReferencedContainerNames(instance.ObjectMeta)
```
Comment (shalberd) [inline components/notebook-controller/controllers/notebook_controller.go:209]: maybe put the two new helper functions in util.go and then refer to `imageChangeTriggerReferencedContainerNames := reconcilehelper.getImageChangeTriggerReferencedContainerNames(instance.ObjectMeta)` ? Minor thing, unsure.

**Comment by shalberd** [comment]
It worked. Tested on an OCP 4.10.43 Cluster without internal Openshift Registry. Used Manifests and Operator Version 1.4.2.
Modified notebook controller image-tag in manifests to test https://github.com/opendatahub-io/odh-manifests/blob/master/odh-notebook-controller/kf-notebook-controller/overlays/openshift/kustomization.yaml#L14

Modified a Notebook CR assembled by odh-dashboard to include the image change trigger annotation. 

Bildschirmfoto 2023-03-21 um 21 57 07

The generated StatefulSet has the same annotation now and the image-field-value is resolved correctly by the image policy admission plug-in from the imagestream name and tag in the annotation to imagestream.tag.from.name, as I do not have an internal openshift registry. 

Bildschirmfoto 2023-03-21 um 20 20 40

Bildschirmfoto 2023-03-21 um 21 59 43

And, most important, the StatefulSet container image field value stays that way :-)

Bildschirmfoto 2023-03-21 um 21 28 40

**Comment by LaVLaS** [comment]
@shalberd Was the original issue caused by the application of `image.openshift.io/triggers` on the `StatefulSet` by k8s and, if so, why does the same annotation need to be applied manually to the `Notebook` CR also

**Comment by shalberd** [comment]
> @shalberd Was the original issue caused by the application of `image.openshift.io/triggers` on the `StatefulSet` by k8s and, if so, why does the same annotation need to be applied manually to the `Notebook` CR also

no, the original issue was not caused by the application of image.openshift.io.triggers on the StatefulSet by K8S. It was caused by the image-field of the container in the notebook PodSpec being overwitten to its original non-resolved value.
The image.openshift.io/triggers annotations fills in the image-field value of the container correctly after the StatefulSet is applied, but then, the notebook controller overwrites it to the original value on reconciling StatefulSet new values and original values. This makes sense, though, because Kubernetes does not know the concept of image trigger admission controllers and imagestreams.

The annotation needs to be applied manually to the Notebook CR also because we do not build StatefulSet CRs, we assemble and submit Notebook CRs, that the contoller reacts to and assembles a StatefulSet from.

Now, here is the historical flow of events leading to this solution:

- ODH Dashboard uses imagestreams to get all the information it needs to assemble a Notebook CR yaml. The kubeflow notebook controller assembles a StatefulSet based on the Notebook CR it finds.
- ODH Dashboard currently puts into the the image-field-value the value of Imagestream status.dockerImageRepository for the image name, while taking the imagestream tag for the image value. Since in environments without internal openshift registry, imagestream status.dockerImageRepository is empty (there is no internal repository), the image-content of the Notebook pod def container is half-empty, missing the image name
- we came to the conclusion that referencing the imagestream name and tag directly via the image.openshift.io.triggers annotation is the way for being independent of the internal openshift registry. We let the openshift image policy admission controller decide whether to reference the internal openshift registry location or the external image location when filling in the container-image-field with the correct value. That triggers annotation works on StatefulSets, Pods, Deployments, DeploymentConfigs, Pods ... any basic Kubernetes object.
- there are basically two parts of this story:
1) ODH Dashboard does not reference the openshift-internal image location directly anymore when assembling the notebook CR. It uses the image trigger annotation. See stories odh-dashboard#792 and odh-dashboard#911 for initial problem and linked PR 800 of ODH dashboard.
2) Notebook Controller still takes all elements and values of the PodSpec when assembling a StatefulSet. However, it disregards overwriting the image-field to its pre-filled-out-value for all containers mentioned in the image-openshift.io/trigers annotation and completed and resolved by the openshift image policy admission plugin.

**Comment by shalberd** [comment]
End-to-end test here:

https://github.com/opendatahub-io/odh-dashboard/pull/800#issuecomment-1481313687

**Comment by VaishnaviHire** [comment]
/hold

Holding this PR until we test a scenario for long running notebooks, and how this change will affect Notebook restarts

**Comment by shalberd** [comment]
> Holding this PR until we test a scenario for long running notebooks, and how this change will affect Notebook restarts

Some thoughts regarding the current situation and how you wish to avoid container restarts during dynamic development / weekly images tag changes:
 
Even in the current implementation with openshift-internal registry and tag, you will have container restarts when the image digest underlying a tag, linked to the imagestream tag, changes. 

The only way you could absolutely watertight avoid container restarts (immutability) when dealing with dynamic images is to reference image digests; one digest per imagestream tag exactly. Have a look at how the status-section of an imagestream changes when the digest of a tagged (imagestream tag) from.name image with tag-format changes on rebuilds. Those SHA256 digests is what openshift works with internally, even outside imagestreams and with regards to imagePullPolicy.

**Comment by VaishnaviHire** [comment]
@shalberd I might be missing some steps, I tested this with `quay.io/opendatahub/kubeflow-notebook-controller:pr-72` image. The changes you added worked. When I add annotation to Notebook instance, it gets added to `Statefulset`
```
kind: StatefulSet
apiVersion: apps/v1
metadata:
  annotations:
    image.openshift.io/triggers: >-
      [{"from":{"kind":"ImageStreamTag","name":"jupyter-minimal-notebook:py3.8-v1",
      "namespace”:”opendatahub”},”fieldPath":"spec.template.spec.containers[?(@.name==\"jupyter-nb-kube-3aadmin\")].image"}]
```

However, the image is not updated 

Podspec
```
      image: >-
        image-registry.openshift-image-registry.svc:5000/opendatahub/jupyter-minimal-notebook:py3.9-v2
```

**Comment by shalberd** [comment]
@VaishnaviHire when you add the annotation to a Notebook CR instance, it is important to check what the name of that first container is. 
`fieldPath":"spec.template.spec.containers[?(@.name==\"jupyter-nb-kube-3aadmin\"`

is the name of the notebook container, the one where you then checked the image-field in the created statefulset, really jupyter-nb-kube-3aadmin? I mean, that name is variable, could be it is jupyter-nb-vaishnavi or something.
In any case, it has to be the same as the container name field value.

**Comment by VaishnaviHire** [comment]
> @VaishnaviHire when you add the annotation manually to a Notebook CR instance, it is important to check what the name of that first container is. `fieldPath":"spec.template.spec.containers[?(@.name==\"jupyter-nb-kube-3aadmin\"`
> 
> is the name of the notebook container, the one where you then checked the image-field in the created statefulset, really `jupyter-nb-kube-3aadmin`? I mean, that name is variable, could be it is `jupyter-nb-vaishnavi` or something`. In ODH Dashboard PR code, the dashboard makes sure that name is exactly as in the containers part of the notebook yaml.
> 
> In any case, it has to be the same as the container name field value. fieldPath in the annotation is kind of a lookup pointing to what image-field to update (of which container, by container name).

Yes it matches with my container name. I also do not see the log message defined [here](https://github.com/opendatahub-io/kubeflow/pull/72/files#diff-ec1007848dc321a9edef149507501c77f59dbbd373efc8b626e4cb26a7d47d09R137).
Re-triggering image build CI job.

**Comment by shalberd** [comment]
@harshad16 this change lets notebook controller reconcile all elements of a notebook / statefulset pod spec when no image change trigger annotation is present. That is, old notebooks assembled the old way will still work as expected.

For newly created notebooks in ODh Dashboard PR 800 way, the image change triggger annotation admission plugin of openshift will handle setting the value of container image field. This change here makes sure that the original placeholder value for the pod container image field from Notebook does not come back to statefulset after image change trigger admission plugin did its work resolving the image-location.

**Comment by shalberd** [comment]
[superseded by PR against v1.7-branch in PR-133](https://github.com/opendatahub-io/kubeflow/pull/133)

**Pull request closed without merging**